coffea/processor/accumulator.py

```diff
@@ -1,11 +1,82 @@
+import copy
+import operator
 from abc import ABCMeta, abstractmethod
-from collections.abc import Set, Mapping
 from collections import defaultdict
+from collections.abc import MutableMapping, MutableSet
+from typing import Dict, Iterable, Optional, Set, TypeVar, Union
+
+try:
+    from typing import Protocol  # type: ignore
+except ImportError:
+    from typing_extensions import Protocol  # type: ignore
+
 import numpy
 
+T = TypeVar("T")
+
+
+class Addable(Protocol):
+    def __add__(self: T, other: T) -> T:
+        ...
+
+
+Accumulatable = Union[Set, Dict, Addable]
+
+
+def add(a: Accumulatable, b: Accumulatable) -> Accumulatable:
+    """Add two accumulatables together, without altering inputs
+
+    This may make copies in certain situations
+    """
+    if isinstance(a, MutableSet) and isinstance(b, MutableSet):
+        return a | b
+    elif isinstance(a, MutableMapping) and isinstance(b, MutableMapping):
+        out = copy.copy(a)
+        out.clear()
+        for key in set(a) & set(b):
+            out[key] = add(a[key], b[key])
+        for key in set(a) - set(b):
+            out[key] = copy.deepcopy(a[key])
+        for key in set(b) - set(a):
+            out[key] = copy.deepcopy(b[key])
+        return out
```

Comment:
It makes more sense in iadd, but do we want add to always take the type of the first operand? I think the usual behavior would be to take the first common non-abstract type shared by a and b (counting back from the passed types)? 

I'm fine with this being the rule so long as we are clear about it. 

What about user defined accumulables which are somehow not compatible with each other? We should detect that and throw an error, to keep people from getting confusing errors.

Reply:
The eventual call to `a + b` would throw if they are not compatible

Reply:
There is a loss of the type of the mapping `b` here but since I figure these are almost always just dictionaries no need to work too hard to resolve the common type of `a` and `b`. It would be a challenging problem. I had originally thought I would convert everything to plain dict in the output, but it might be nice for user to have some chance of receiving what they put in (e.g. defaultdict)

Reply:
Another option is to require `type(a) == type(b)` which in this context is how we would expect things to be used regardless

Reply:
I went for a bit of a softer type check, take a look at the tests to see how it plays out

Reply:
That's also a fine solution. I think what we have right now is not confusing.